src/bun.js/bindings/NodeVMSyntheticModule.cpp

```diff
@@ -0,0 +1,232 @@
+#include "NodeVMSourceTextModule.h"
+#include "NodeVMSyntheticModule.h"
+
+#include "AsyncContextFrame.h"
+#include "ErrorCode.h"
+#include "JSDOMExceptionHandling.h"
+
+#include "wtf/Scope.h"
+
+#include "JavaScriptCore/JIT.h"
+#include "JavaScriptCore/JSModuleEnvironment.h"
+#include "JavaScriptCore/JSModuleRecord.h"
+#include "JavaScriptCore/JSPromise.h"
+#include "JavaScriptCore/JSSourceCode.h"
+#include "JavaScriptCore/ModuleAnalyzer.h"
+#include "JavaScriptCore/ModuleProgramCodeBlock.h"
+#include "JavaScriptCore/Parser.h"
+#include "JavaScriptCore/SourceCodeKey.h"
+#include "JavaScriptCore/Watchdog.h"
+
+#include "../vm/SigintWatcher.h"
+
+namespace Bun {
+using namespace NodeVM;
+
+NodeVMSyntheticModule* NodeVMSyntheticModule::create(VM& vm, JSGlobalObject* globalObject, ArgList args)
+{
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue identifierValue = args.at(0);
+    if (!identifierValue.isString()) {
+        throwArgumentTypeError(*globalObject, scope, 0, "identifier"_s, "Module"_s, "Module"_s, "string"_s);
+        return nullptr;
+    }
+
+    JSValue contextValue = args.at(1);
+    if (contextValue.isUndefined()) {
+        contextValue = globalObject;
+    } else if (!contextValue.isObject()) {
+        throwArgumentTypeError(*globalObject, scope, 1, "context"_s, "Module"_s, "Module"_s, "object"_s);
+        return nullptr;
+    }
+
+    JSValue exportNamesValue = args.at(2);
+    auto* exportNamesArray = jsDynamicCast<JSArray*>(exportNamesValue);
+    if (!exportNamesArray) {
+        throwArgumentTypeError(*globalObject, scope, 2, "exportNames"_s, "Module"_s, "Module"_s, "Array"_s);
+        return nullptr;
+    }
+
+    JSValue syntheticEvaluationStepsValue = args.at(3);
+    if (!syntheticEvaluationStepsValue.isUndefined()) {
+        if (!syntheticEvaluationStepsValue.isCallable()) {
+            throwArgumentTypeError(*globalObject, scope, 3, "syntheticEvaluationSteps"_s, "Module"_s, "Module"_s, "function"_s);
+            return nullptr;
+        }
+
+        syntheticEvaluationStepsValue = AsyncContextFrame::withAsyncContextIfNeeded(globalObject, syntheticEvaluationStepsValue);
+    }
+
+    JSValue moduleWrapperValue = args.at(4);
+    if (!moduleWrapperValue.isObject()) {
+        throwArgumentTypeError(*globalObject, scope, 4, "moduleWrapper"_s, "Module"_s, "Module"_s, "object"_s);
+        return nullptr;
+    }
+
+    WTF::HashSet<String> exportNames;
+    for (unsigned i = 0; i < exportNamesArray->getArrayLength(); i++) {
+        JSValue exportNameValue = exportNamesArray->getDirectIndex(globalObject, i);
```

Comment:
i think we want to use `getIndex` instead of `getDirectIndex` because these could be getters